tpm2/direct/reflect.go

```diff
@@ -0,0 +1,999 @@
+// Package direct provides 1:1 mapping to TPM 2.0 APIs.
+package direct
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-tpm/tpmutil"
+)
+
+const (
+	// Chosen based on MAX_DIGEST_BUFFER, the length of the longest
+	// reasonable list returned by the reference implementation.
+	maxListLength uint32 = 1024
+)
+
+// TPM represents a logical connection to a TPM. Support for
+// various commands is provided at runtime using reflection.
+type TPM struct {
+	transport io.ReadWriteCloser
+}
+
+// NewTPM opens a TPM connection using the provided ReadWriteCloser.
+// When this TPM connection is closed, the transport is closed.
+func NewTPM(t io.ReadWriteCloser) *TPM {
+	return &TPM{
+		transport: t,
+	}
+}
+
+// Close closes the connection to the TPM and its underlying
+// transport.
+func (t *TPM) Close() error {
+	return t.transport.Close()
+}
+
+// Execute sends the provided command and returns the TPM's response.
+func (t *TPM) Execute(cmd Command, rsp Response, extraSess ...Session) error {
+	cc := cmd.Command()
+	if rsp.Response() != cc {
+		return fmt.Errorf("cmd and rsp must be for same command: %v != %v", cc, rsp.Response())
+	}
+	sess, err := cmdAuths(cmd)
+	if err != nil {
+		return err
+	}
+	sess = append(sess, extraSess...)
+	if len(sess) > 3 {
+		return fmt.Errorf("too many sessions: %v", len(sess))
+	}
+	hasSessions := len(sess) > 0
+	// Initialize the sessions, if needed
+	for i, s := range sess {
+		if err := s.Init(t); err != nil {
+			return fmt.Errorf("initializing session %d: %w", i, err)
+		}
+		if err := s.NewNonceCaller(); err != nil {
+			return err
+		}
+	}
+	handles := cmdHandles(cmd)
+	names, err := cmdNames(cmd)
+	if err != nil {
+		return err
+	}
+	parms, err := cmdParameters(cmd, sess)
+	if err != nil {
+		return err
+	}
+	sessions, err := cmdSessions(t, sess, cc, names, parms)
+	if err != nil {
+		return err
+	}
+	hdr := cmdHeader(hasSessions, 10 /* size of command header */ +len(handles)+len(sessions)+len(parms), cc)
+	command := append(hdr, handles...)
+	command = append(command, sessions...)
+	command = append(command, parms...)
+
+	// Send the command via the transport.
+	response, err := tpmutil.RunCommandRaw(t.transport, command)
+	if err != nil {
+		return err
+	}
+
+	// Parse the command directly into the response structure.
+	rspBuf := bytes.NewBuffer(response)
+	err = rspHeader(rspBuf)
+	if err != nil {
+		var bonusErrs []string
+		// Emergency cleanup, then return.
+		for _, s := range sess {
+			if err := s.CleanupFailure(t); err != nil {
+				bonusErrs = append(bonusErrs, err.Error())
+			}
+		}
+		if len(bonusErrs) != 0 {
+			return fmt.Errorf("%w - additional errors encountered during cleanup: %v", err, strings.Join(bonusErrs, ", "))
```

Comment:
Consider using MultiErr

Reply:
Don't want to add a dependency for this one thing, where there is clearly a "main" error and then some extra info about other things that failed.